keps/sig-node/4010-head-request-probes/README.md

```diff
@@ -0,0 +1,264 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This Kubernetes Enhancement Proposal (KEP) introduces the capability to use HTTP HEAD requests for defining readiness and liveness probes in Kubernetes pods. Currently, only GET requests are supported for probes, limiting the options for checking the health of containers. By allowing the use of HEAD requests, Kubernetes users will have more flexibility and finer-grained control over their probes, especially for customer-facing applications.
+
+## Motivation
+
+The motivation behind this proposal is to enhance Kubernetes' readiness and liveness probes by supporting HTTP HEAD requests. While GET requests are commonly used, they might not always be suitable for certain scenarios, such as checking the availability of resources without fetching their content. The ability to use HEAD requests will enable more efficient health checks for specific use cases, particularly those involving customer-facing services. By extending the probe functionality, Kubernetes aims to provide a more versatile and customizable platform for managing container health.
+
+### Goals
+
+- Allow Kubernetes users to utilize HTTP HEAD requests for defining readiness and liveness probes.
+- Enhance the probe functionality to support more efficient and lightweight health checks.
+- Provide additional flexibility for customer-facing applications that require specific health check behaviors.
+
+### Non-Goals
+
+- Modifying existing HTTP GET-based probes.
+- Introducing new HTTP request methods beyond HEAD.
+
+## Proposal
+
+The proposal involves extending Kubernetes' readiness and liveness probes to support HTTP HEAD requests. This will involve modifying the Kubernetes API to allow users to specify the request method as HEAD for these probes. When a HEAD request is used, only the response headers will be retrieved, without the response body. This lightweight operation is ideal for checking the availability of resources without incurring unnecessary network traffic or consuming additional system resources.
+
+For example, a pod's readiness probe configuration can be updated as follows:
+
+```yaml
+readinessProbe:
+  httpHead:
+    headers:
+      - name: X-Custom-Header
+        value: some-value
+    queryParams:
+      - name: token
+        value: abc123
+    path: /health
+    port: 8080
+    scheme: HTTPS
+    timeoutSeconds: 3
+    periodSeconds: 5
+```
+
+In this example, the readiness probe is configured to use an HTTP HEAD request with custom headers, query parameters, and a specific path, port, and scheme.
+
+There is an [untested solution how this feature might be implemented](https://github.com/kubernetes/kubernetes/commit/e3d556f0920ff736fec5df73c42f6db93f436a37), without the usage of a feature gate.
```

Comment:
I think there should be a note that a feature gate (the one listed below, perhaps) will be needed in the final implementation

Reply:
Good point, I will add something like

```suggestion
There is an [untested solution how this feature might be implemented](https://github.com/kubernetes/kubernetes/commit/e3d556f0920ff736fec5df73c42f6db93f436a37).  
We need to ensure that a feature gate is added in the final implementation.
```

WDYT?

Reply:
That works! The wording could also be
`Note: this solution is also missing a feature gate, which will be required for the final implementation`
I'm open either way